packages/plugins/typescript/operations/tests/ts-documents.spec.ts

```diff
@@ -5560,6 +5559,58 @@ export type KittyQuery = { __typename?: 'Query', animals: Array<{ __typename?: '
         );
       `);
     });
+
+    it('On avoidOptionals:true, optionals (?) on types should be avoided', async () => {
+      const schema = buildSchema(/* GraphQL */ `
+        type Query {
+          me: User!
+        }
+
+        type User {
+          messages: [Message!]!
+        }
+
+        type Message {
+          content: String!
+        }
+      `);
+
+      const fragment = parse(/* GraphQL */ `
+        query MyQuery($include: Boolean!) {
+          me {
+            messages @include(if: $include) {
+              content
+            }
+          }
+        }
+      `);
+
+      const { content } = await plugin(
+        schema,
+        [{ location: '', document: fragment }],
+        {
+          avoidOptionals: true,
+          nonOptionalTypename: true,
+          preResolveTypes: false,
+        },
+        {
+          outputFile: 'graphql.ts',
+        }
+      );
+
+      expect(content).toBeSimilarStringTo(`
+        export type MyQueryQuery = (
+          { __typename: 'Query' }
+          & { me: (
+            { __typename: 'User' }
+            & { messages: Maybe<Array<(
+              { __typename: 'Message' }
+              & Pick<Message, 'content'>
+            )>> }
```

Comment:
This should actually generate the following:

```suggestion
            & { messages?: Array<(
              { __typename: 'Message' }
              & Pick<Message, 'content'>
            )> }
```

which is what the actual server response would be

Reply:
discussed [here](https://github.com/dotansimha/graphql-code-generator/issues/6748#issuecomment-931123054)